File: go/007.go
package main

type Primer struct {
	cur int64
	primes []int64
}

func (primer *Primer) nextPrime() int64 {
	isPrime := func(n int64) bool {
		for _, prime := range primer.primes {
			if n == prime {
				return true
			}
			if n % prime == 0 {
				return false
			}
		}
		return true
	}

	for primer.cur++; !isPrime(primer.cur); primer.cur++ {}
	primer.primes = append(primer.primes, primer.cur)

	return primer.cur
}

func (primer *Primer) reset() {
	primer.cur = 1
}

func main() {
	var primer Primer
	primer.reset()
	for i := 1; i < 10001; i++ { primer.nextPrime() }
	println(primer.nextPrime())
}

File: r/007.r
#!/usr/bin/Rscript

nextPrime <- (function() {
    cur <- 1
    primes <- NULL

    isNextPrime <- function(n) {
        for (prime in primes) {
            if (n %% prime == 0) {
                return(FALSE)
            }
        }
        TRUE
    }

    function() {
        cur <<- cur + 1
        while (!isNextPrime(cur)) {
            cur <<- cur + 1
        }

        primes <<- append(primes, cur)

        cur
    }
})()

for (i in 1:1e4) {
    nextPrime()
}

cat(nextPrime(), "\n")
